bundles/org.openhab.binding.entsoe/src/main/java/org/openhab/binding/entsoe/internal/EntsoeHandler.java

```diff
@@ -0,0 +1,272 @@
+ * sent to one of the channels.
+ *
+ * @author Jørgen Melhus - Initial contribution
+ */
+@NonNullByDefault
+public class EntsoeHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EntsoeHandler.class);
+
+    private EntsoeConfiguration config;
+
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    private Unit<Currency> baseUnit = CurrencyUnits.BASE_CURRENCY;
+    private Unit<Currency> fromUnit = new CurrencyUnit(EntsoeBindingConstants.ENTSOE_CURRENCY, null);
+
+    private Map<ZonedDateTime, Double> responseMap;
+
+    // Use region-based ZoneId instead of CET to handle transition to CEST automatically
+    private ZonedDateTime lastDayAheadReceived = ZonedDateTime.of(LocalDateTime.MIN, ZoneId.of("Europe/Paris"));
+
+    private int historicDaysInitially = 0;
+
+    public EntsoeHandler(Thing thing) {
+        super(thing);
+        config = getConfigAs(EntsoeConfiguration.class);
+        responseMap = new TreeMap<>();
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        logger.trace("channelLinked(channelUID:{})", channelUID.getAsString());
+        String channelID = channelUID.getId();
+
+        if (EntsoeBindingConstants.CHANNEL_SPOT_PRICE.equals(channelID)) {
+            if (responseMap.isEmpty()) {
+                refreshPrices();
+            }
+            updateCurrentHourState(EntsoeBindingConstants.CHANNEL_SPOT_PRICE);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> refreshJob = this.refreshJob;
+        if (refreshJob != null) {
+            refreshJob.cancel(true);
+            this.refreshJob = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.trace("handleCommand(channelUID:{}, command:{})", channelUID.getAsString(), command.toFullString());
+
+        if (command instanceof RefreshType) {
+            refreshPrices();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(EntsoeConfiguration.class);
+
+        if (historicDaysInitially == 0) {
+            historicDaysInitially = config.historicDays;
+        }
+        updateStatus(ThingStatus.UNKNOWN);
+        refreshJob = scheduler.schedule(this::refreshPrices, 0, TimeUnit.SECONDS);
+    }
+
+    private ZonedDateTime currentCetTime() {
+        // Use region-based ZoneId instead of CET to handle transition to CEST automatically
+        return ZonedDateTime.now(ZoneId.of("Europe/Paris"));
+    }
+
+    private ZonedDateTime currentCetTimeWholeHours() {
+        return currentCetTime().truncatedTo(ChronoUnit.HOURS);
+    }
+
+    private BigDecimal getCurrentHourKwhPrice() throws EntsoeResponseMapException {
+        if (responseMap.isEmpty()) {
+            throw new EntsoeResponseMapException("responseMap is empty");
+        }
+        var currentCetTime = currentCetTime();
```

Comment:
This leads to the wrong results. The prices that you are fetching are in UTC. The current hour must also be calculated with UTC time-zone. Sample: If it is now 21:00 Europe / Paris then this would be 19:00 UTC. But you are doing the lookup with 21. That means we are getting the prices for 23:00 Europe / Paris.

Reply:
Perhaps it would be simpler having `Instant` as key in the response map then, rather than `ZonedDateTime`?

Reply:
When updating current hour state I'm comparing hour and day of entsoe values with current hour and day, I think it is more convenient to keep as zoneddatetime and convert to instant when needed.

Reply:
I will need to have a closer at look to understand the logic behind comparing hour without considering date (i.e. why hour is compared rather than full `Instant`/UTC timestamp).

Reply:
I had another look. Maybe I'm missing something, but wouldn't it be even simpler to just compare `Instant` directly? Given that `EntsoeTimeSerie.getTime()` would return an `Instant`:
```java
        Instant currentHourStart = Instant.now().truncatedTo(ChronoUnit.HOURS);
        var result = entsoeTimeSeries.stream().filter(x -> x.getTime().equals(currentHourStart)).findFirst();
```

Reply:
yeah, probably simpler, changed it to instant